mycpp/gc_str.cc

```diff
@@ -298,12 +303,20 @@ BigStr* BigStr::replace(BigStr* old, BigStr* new_str) {
 
   // First pass: Calculate number of replacements, and hence new length
   int replace_count = 0;
-  while (p_this <= last_possible) {
-    if (memcmp(p_this, old_data, old_len) == 0) {  // equal
-      replace_count++;
-      p_this += old_len;
-    } else {
-      p_this++;
+  if (old_len == 0) {
```

Comment:
why do we need to special case old_len == 0 ?

OK this is a little weird and I honestly wonder if the code was correct

```
>>> 'xxx'.replace('', 'y', 3)
'yxyxyx'

```

I think we should check if there are unit tests in `gc_str_test.cc` that cover this

Whenever we add more if statements in the C++, I think we should have more C++ unit tests

because it's not obvious if the spec tests cover them

there could have been a bug here perhaps

Reply:
I tried altering the previous string replacement algorithm to support `old_len == 0`, however it ended up so full of special cases that I decided to handle `old_len` separately.

Previously the code would not terminate on an empty string. I opted to follow the python2 behaviour here. However, the `Str => replace()` in YSH could behave differently. Did you have any thoughts there?

Re unit tests: sure, I can write some. (By the way, I did sneak in some spec tests which cover this behaviour. See the ones with empty strings and/or counts.)

Reply:
Oh I didn't realize you fixed a bug!  That's important, we should note it in the "change description"

Since I only look at the  whole diff, it's usually OK to squash and then write a PR description

That will help review the changes

---

OK I didn't see the spec tests, but I don't think it's a bad idea to simply copy them in to `gc_str_test.cc`, since it will help us debug ...

I'm still kinda thinking about empty string, even though it is consistent between JS and Python

Reply:
BTW I use `devtools/git.sh rebase-staging`, although the command it runs is actually shorter :-P

https://github.com/oilshell/oil/blob/master/devtools/git.sh#L30

It's not essential to do that every time, because when I merge PRs I squash

But it can be useful to consolidate things under a single big change description.  This PR can use that, if there is both a bug fix and a new feature